### PR TITLE
Use outputs for model output names if outputs are dict

### PR DESCRIPTION
**Pull request by eugen-vusak**
PR related to #16758

### CODE REVIEW

**Comment by gbaned** [comment]
@eugen-vusak Can you please sign CLA. Thank you!

**Review by fchollet — COMMENTED** [review]
Thanks for the PR.

First of all, anything we do for a dict output should also be applied for a dict input. They should be consistent.

Second, when you create a Functional model with a dict input or output, there is some ambiguity, because there are two sets of output names: the keys in the dict, and the names of the layers the inputs or outputs came from.

Picking one over the other does not resolve the ambiguity. Changing which one we pick would also be a breaking change, which we cannot do.

Instead, I recommend introducing a check of whether the layer names and the keys are the same, and if they aren't, printing a warning asking the user to ensure they are the same.

Keeping both sets of names identical is the only way to clear up the ambiguity.

**Comment by eugen-vusak** [comment]
Sure, I agree with everything said, 

Problem is that with inputs user can set names, using name argument in the `InputLayer`, but for outputs user does not have any control over how outputs will be named. Only solution I see is to use `tf.identity` for every output in order do gave it appropriate name. 

This is important because, for example, when exporting API, output should stay the same regardless of implementation details. Or name of the loss that is also dependent on the output names should be consistent because of logging and monitoring. 

Do you maybe have different solution in mind? For example adding additional argument for output_names?

Maybe make `tf.keras.layers.Layer` name it's outputs based on return dict just like tf.keras.Model does? For example we have layer called `my_layer` and it's `call` function returns {"a": a, "b": b}, it's outputs will be called `my_layer_0` and `my_layer_1`, maybe they should be call `my_layer_a` and `my_layer_b`?

**Comment by fchollet** [comment]
> Maybe make tf.keras.layers.Layer name it's outputs based on return dict just like tf.keras.Model does? For example we have layer called my_layer and it's call function returns {"a": a, "b": b}, it's outputs will be called my_layer_0 and my_layer_1, maybe they should be call my_layer_a and my_layer_b?

I'm not sure I get the proposal -- outputs are tensors and thus aren't named by the framework. Tensor names are an internal implementation detail, not an API. They're autogenerated.

**Comment by gbaned** [comment]
@eugen-vusak Can you please check @fchollet's comments and keep us posted ? Thank you!
